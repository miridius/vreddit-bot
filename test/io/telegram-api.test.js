const { CHAT } = require('../helpers');
const telegramApi = require('../../src/io/telegram-api');
const { existsSync, copyFileSync } = require('fs');
const nock = require('nock');
const filenamify = require('filenamify');

nock.back.fixtures = __dirname + '/__fixtures__/';
nock.back.setMode(process.env.CI ? 'lockdown' : 'record');

// Use nock.back to mock all HTTP requests (if any) for each test
beforeEach(async () => {
  const state = expect.getState();
  state.nockBack = await nock.back(`${filenamify(state.currentTestName)}.json`);
});

afterEach(() => {
  const { nockBack } = expect.getState();
  nockBack.nockDone();
  nockBack.context.assertScopesFinished();
});

afterAll(nock.restore);

// Since form boundary is generated randomly we need to make it deterministic
Math.random = jest.fn(() => 0.5);

describe('telegramApi', () => {
  it('works with no params (getMe)', () => {
    return expect(telegramApi('getMe')).resolves.toMatchSnapshot();
  });

  it('works with normal params (sendMessage)', () => {
    const method = 'sendMessage';
    const text = 'Test Message';
    const params = {
      chat_id: CHAT.id,
      text,
      reply_to_message_id: 35,
      reply_markup: {
        inline_keyboard: [[{ text: 'Click me', url: 'https://example.com' }]],
      },
    };
    return expect(telegramApi(method, params)).resolves.toMatchSnapshot();
  });

  it('works with file params (sendVideo)', () => {
    const video = `${__dirname}/__fixtures__/hf352syjjka61.mp4`;
    if (!existsSync(video)) copyFileSync(video + '.save', video);
    const method = 'sendVideo';
    const params = {
      chat_id: CHAT.id,
      width: 406,
      height: 720,
      reply_to_message_id: undefined,
    };
    return expect(
      telegramApi(method, params, { video }),
    ).resolves.toMatchSnapshot();
  });

  it('throws an error if result is not ok', () => {
    return expect(telegramApi).rejects.toThrowErrorMatchingInlineSnapshot(
      `"Telegram API error: Unauthorized"`,
    );
  });
});
