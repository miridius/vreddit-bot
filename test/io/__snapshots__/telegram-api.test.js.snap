// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`telegramApi works with file params (sendVideo) 1`] = `
Object {
  "ok": true,
  "result": Object {
    "chat": Object {
      "first_name": "Dave",
      "id": 123456,
      "last_name": "Rolle",
      "type": "private",
      "username": "DavidRolle",
    },
    "date": 1611448894,
    "from": Object {
      "first_name": "v.redd.it",
      "id": 54321,
      "is_bot": true,
      "username": "vreddit_bot",
    },
    "message_id": 44,
    "video": Object {
      "duration": 11,
      "file_id": "BAACAgIAAxkDAAMsYAzCPc1oRx1yMPNTMvQbq0UIYFcAAqcKAAJavGlI67jMXUiP1L0eBA",
      "file_name": "hf352syjjka61.mp4",
      "file_size": 3264564,
      "file_unique_id": "AgADpwoAAlq8aUg",
      "height": 720,
      "mime_type": "video/mp4",
      "thumb": Object {
        "file_id": "AAMCAgADGQMAAyxgDMI9zWhHHXIw81My9BurRQhgVwACpwoAAlq8aUjruMxdSI_UvV9pvZouAAMBAAdtAANzdQACHgQ",
        "file_size": 10120,
        "file_unique_id": "AQADX2m9mi4AA3N1AAI",
        "height": 320,
        "width": 180,
      },
      "width": 406,
    },
  },
}
`;

exports[`telegramApi works with no params (getMe) 1`] = `
Object {
  "ok": true,
  "result": Object {
    "can_join_groups": true,
    "can_read_all_group_messages": true,
    "first_name": "v.redd.it",
    "id": 54321,
    "is_bot": true,
    "supports_inline_queries": true,
    "username": "vreddit_bot",
  },
}
`;

exports[`telegramApi works with normal params (sendMessage) 1`] = `
Object {
  "ok": true,
  "result": Object {
    "chat": Object {
      "first_name": "Dave",
      "id": 123456,
      "last_name": "Rolle",
      "type": "private",
      "username": "DavidRolle",
    },
    "date": 1611692688,
    "from": Object {
      "first_name": "v.redd.it",
      "id": 54321,
      "is_bot": true,
      "username": "vreddit_bot",
    },
    "message_id": 54,
    "reply_markup": Object {
      "inline_keyboard": Array [
        Array [
          Object {
            "text": "Click me",
            "url": "https://example.com/",
          },
        ],
      ],
    },
    "reply_to_message": Object {
      "chat": Object {
        "first_name": "Dave",
        "id": 123456,
        "last_name": "Rolle",
        "type": "private",
        "username": "DavidRolle",
      },
      "date": 1611414446,
      "entities": Array [
        Object {
          "length": 31,
          "offset": 0,
          "type": "url",
        },
      ],
      "from": Object {
        "first_name": "Dave",
        "id": 123456,
        "is_bot": false,
        "language_code": "en",
        "last_name": "Rolle",
        "username": "DavidRolle",
      },
      "message_id": 35,
      "text": "https://v.redd.it/rk2repuuvc261",
    },
    "text": "Test Message",
  },
}
`;
